python/delta/tests/test_deltatable.py

```diff
@@ -912,6 +912,26 @@ def optimize() -> None:
                          "Predicate references non-partition column 'value'. "
                          "Only the partition columns may be referenced: [key]")
 
+        def test_optimize_zorder(self) -> None:
+            # write an unoptimized delta table
+            df = self.spark.createDataFrame([("a", 1), ("a", 2)], ["key", "value"]).repartition(1)
+            df.write.format("delta").save(self.tempFile)
+            df = self.spark.createDataFrame([("a", 3), ("a", 4)], ["key", "value"]).repartition(1)
+            df.write.format("delta").save(self.tempFile, mode="append")
+            df = self.spark.createDataFrame([("b", 1), ("b", 2)], ["key", "value"]).repartition(1)
+            df.write.format("delta").save(self.tempFile, mode="append")
+
+            # create DeltaTable
+            dt = DeltaTable.forPath(self.spark, self.tempFile)
+
+            # execute Z-Order Optimization
+            optimizer = dt.optimize()
+            res = optimizer.executeZOrderBy(["a", "b"])
+
+            # assertions
+            self.assertTrue(isinstance(optimizer, DeltaOptimizeBuilder))
+            self.assertTrue(isinstance(res, DataFrame))
```

Comment:
This is a good starting point for tests. You can expand this test to verify that the returned DataFrame contains the expected values. Given there are 3 files in the input, we expect a 1 one file as output (as the total size is less than the maxFileSize) (Example verification for optimize compaction is [here](https://github.com/delta-io/delta/blob/master/python/delta/tests/test_deltatable.py#L853)). And we need to verify that there are zOrder metrics (Scala based tests for verification are [here](https://github.com/delta-io/delta/blob/master/core/src/test/scala/org/apache/spark/sql/delta/optimize/OptimizeMetricsSuite.scala#L236)). You could actually copy [this](https://github.com/delta-io/delta/blob/master/core/src/test/scala/org/apache/spark/sql/delta/optimize/OptimizeMetricsSuite.scala#L236) test.

Additional tests needed:
- test that does ZOrder By and also contains WHERE clause.
- a negative test. You could write a test where we pass a partition column as z-order by columns. Scala version of the test is [here](https://github.com/delta-io/delta/blob/master/core/src/test/scala/org/apache/spark/sql/delta/optimize/OptimizeZOrderSuite.scala#L58).